com.ca.lsp.cobol/lsp-core-cobol-parser/src/main/java/com/ca/lsp/core/cobol/preprocessor/sub/document/impl/CobolSemanticParserListenerImpl.java

```diff
@@ -0,0 +1,292 @@
+  private final SemanticContext semanticContext;
+  private final PreprocessorListener listener;
+
+  CobolSemanticParserListenerImpl(
+      final BufferedTokenStream tokens,
+      final SemanticContext semanticContext,
+      final CobolPreprocessor.CobolSourceFormatEnum format,
+      PreprocessorListener listener) {
+    this.tokens = tokens;
+    this.format = format;
+    this.semanticContext = semanticContext;
+    this.listener = listener;
+
+    contexts.push(new CobolDocumentContext());
+  }
+
+  private final PreprocessorCleanerServiceImpl preprocessorCleanerService =
+      new PreprocessorCleanerServiceImpl(contexts);
+
+  @Override
+  public CobolDocumentContext context() {
+    return contexts.peek();
+  }
+
+  @Override
+  public void enterCompilerOptions(final CobolPreprocessorParser.CompilerOptionsContext ctx) {
+    // push a new context for COMPILER OPTIONS terminals
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void enterReplaceArea(final CobolPreprocessorParser.ReplaceAreaContext ctx) {
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void enterReplaceByStatement(final CobolPreprocessorParser.ReplaceByStatementContext ctx) {
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void enterReplaceOffStatement(
+      final CobolPreprocessorParser.ReplaceOffStatementContext ctx) {
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void enterTitleStatement(final CobolPreprocessorParser.TitleStatementContext ctx) {
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void enterParagraphName(final CobolPreprocessorParser.ParagraphNameContext ctx) {
+    semanticContext.getParagraphs().define(ctx.getText().toUpperCase(), retrievePosition(ctx));
+  }
+
+  @Override
+  public void enterCopyStatement(final CobolPreprocessorParser.CopyStatementContext ctx) {
+    this.preprocessorCleanerService.push();
+  }
+
+  @Override
+  public void exitCompilerOptions(final CobolPreprocessorParser.CompilerOptionsContext ctx) {
+    // throw away COMPILER OPTIONS terminals
+    this.preprocessorCleanerService.pop();
+  }
+
+  @Override
+  public void exitDataDescriptionEntryFormat1(
+      CobolPreprocessorParser.DataDescriptionEntryFormat1Context ctx) {
+    createVariableAndDefine(ctx);
+  }
+
+  @Override
+  public void exitDataDescriptionEntryFormat2(
+      CobolPreprocessorParser.DataDescriptionEntryFormat2Context ctx) {
+    createVariableAndDefine(ctx);
+  }
+
+  @Override
+  public void exitDataDescriptionEntryFormat3(
+      CobolPreprocessorParser.DataDescriptionEntryFormat3Context ctx) {
+    createVariableAndDefine(ctx);
+  }
+
+  private void createVariableAndDefine(ParserRuleContext ctx) {
+    String levelNumber;
+
+    if (ctx instanceof CobolPreprocessorParser.DataDescriptionEntryFormat1Context) {
+      CobolPreprocessorParser.DataDescriptionEntryFormat1Context ctxDataDescF1 =
```

Comment:
you can not skip the cast by attributing the value directly to the variables?

Reply:
I will assign to myself as task to perform next week